src/libAtomVM/ets.c

```diff
@@ -388,3 +420,117 @@ EtsErrorCode ets_delete(term ref, term key, term *ret, Context *ctx)
 
     return EtsOk;
 }
+
+bool operation_to_tuple4(term operation, term *position, term *increment, term *threshold, term *set_value)
+{
+    if (term_is_integer(operation)) {
+        *increment = operation;
+        *position = term_from_int(2);
+        *threshold = term_invalid_term();
+        *set_value = term_invalid_term();
+        return true;
+    }
+
+    if (!term_is_tuple(operation)) {
+        return false;
+    }
+    int n = term_get_tuple_arity(operation);
+    if (n != 2 && n != 4) {
+        return false;
+    }
+
+    term pos = term_get_tuple_element(operation, 0);
+    term incr = term_get_tuple_element(operation, 1);
+    if (!term_is_integer(pos) || !term_is_integer(incr)) {
+        return false;
+    }
+
+    if (n == 2) {
+        *position = pos;
+        *increment = incr;
+        *threshold = term_invalid_term();
+        *set_value = term_invalid_term();
+        return true;
+    }
+
+    term tresh = term_get_tuple_element(operation, 2);
+    term set_val = term_get_tuple_element(operation, 3);
+    if (!term_is_integer(tresh) || !term_is_integer(set_val)) {
+        return false;
+    }
+
+    *position = pos;
+    *increment = incr;
+    *threshold = tresh;
+    *set_value = set_val;
+    return true;
+}
+
+EtsErrorCode ets_update_counter(term ref, term key, term operation, term default_value, term *ret, Context *ctx)
+{
+    struct EtsTable *ets_table = term_is_atom(ref) ? ets_get_table_by_name(&ctx->global->ets, ref, TableAccessWrite) : ets_get_table_by_ref(&ctx->global->ets, term_to_ref_ticks(ref), TableAccessWrite);
+    if (ets_table == NULL) {
+        return EtsTableNotFound;
+    }
+    term to_insert = term_invalid_term();
+    term list = term_invalid_term();
+    EtsErrorCode result = ets_table_lookup(ets_table, key, &list, ctx);
+    if (result != EtsOk) {
+        SMP_UNLOCK(ets_table);
+        return result;
+    }
+    if (term_is_nil(list)) {
+        if (term_is_invalid_term(default_value)) {
+            SMP_UNLOCK(ets_table);
+            return EtsBadEntry;
+        }
+        to_insert = default_value;
+    } else {
+        to_insert = term_get_list_head(list);
+    }
+
+    if (!(term_is_tuple(to_insert))) {
+        SMP_UNLOCK(ets_table);
+        return EtsBadEntry;
+    }
+    term position_term, increment_term, threshold_term, set_value_term;
+
+    if (!operation_to_tuple4(operation, &position_term, &increment_term, &threshold_term, &set_value_term)) {
+        SMP_UNLOCK(ets_table);
+        return EtsBadEntry;
+    }
+    int arity = term_get_tuple_arity(to_insert);
+    int position = term_to_int(position_term) - 1;
+    if (arity <= position || position < 1) {
+        SMP_UNLOCK(ets_table);
+        return EtsBadEntry;
+    }
+
+    term elem = term_get_tuple_element(to_insert, position);
+    if (!term_is_integer(elem)) {
+        SMP_UNLOCK(ets_table);
+        return EtsBadEntry;
+    }
+    int increment = term_to_int(increment_term);
+    // We don't check overflow here.
```

Comment:
I'm not a really huge fan of this, we have a lot of simple to use helpers for providing overflow handling.
See also: `src/libAtomVM/overflow_helpers.h` and bif.c:
```c
if (BUILTIN_ADD_OVERFLOW_INT(val1, val2, &res)) {
    // handle overflow here
}
```

Also we might investigate if using a 64 bit integer might be useful.